specs/experimental/op-stack-manager.md

```diff
@@ -0,0 +1,261 @@
+[contract releases]: https://github.com/ethereum-optimism/optimism/blob/main/packages/contracts-bedrock/VERSIONING.md
+[interop specification]: ../interop/overview.md
+[ethereum-optimism/specs#94]: https://github.com/ethereum-optimism/specs/issues/94
+[admin roles]: https://specs.optimism.io/protocol/configurability.html#admin-roles
+[`SystemConfig` contract]: https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.3.0/packages/contracts-bedrock/src/L1/SystemConfig.sol
+[`L2OutputOracle` contract]: https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.3.0/packages/contracts-bedrock/src/L1/L2OutputOracle.sol
+[standard configuration]: ../protocol/configurability.md
+[superchain registry]: https://github.com/ethereum-optimism/superchain-registry
+[ethereum-lists/chains]: https://github.com/ethereum-lists/chains
+[Batch Inbox]: ../protocol/configurability.md#consensus-parameters
+
+The OP Stack Manager is a contract that deploys the L1 contracts for an OP Stack chain in a single
+transaction. It provides a minimal set of user-configurable parameters to ensure that the resulting
+chain meets the [standard configuration] requirements.
+
+The version deployed is always a governance-approved contract release. The set
+of governance approved [contract releases] can be found on the
+[Optimism Monorepo releases] page, and is the set of releases named
+`op-contracts/vX.Y.Z`.
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+**Table of Contents**
+
+- [OP Stack Manager](#op-stack-manager)
+  - [Deployment](#deployment)
+  - [Interface](#interface)
+    - [`deploy`](#deploy)
+    - [Getter Methods](#getter-methods)
+  - [Implementation](#implementation)
+    - [Batch Inbox Address](#batch-inbox-address)
+    - [Contract Deployments](#contract-deployments)
+  - [Security Considerations](#security-considerations)
+    - [Chain ID Source of Truth](#chain-id-source-of-truth)
+    - [Chain ID Frontrunning](#chain-id-frontrunning)
+    - [Chain ID Value](#chain-id-value)
+    - [Proxy Admin Owner](#proxy-admin-owner)
+    - [Upgradeability (ABI Changes)](#upgradeability-abi-changes)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Deployment
+
+The OP Stack Manager is deployed at `0xTODO`. It can be deployed as follows:
+
+TODO.
+
+## Interface
+
+Version 1.0.0 of the OP Stack Manager deploys the [`op-contracts/v1.3.0`]
+contracts release.
+
+### `deploy`
+
+The `deploy` method is the only non-view method in the contract. It is used to
+deploy the full set of L1 contracts required to setup a new OP Stack chain that
+complies with the [standard configuration]. It has the following interface:
+
+```solidity
+function deploy(
+  uint64 l2ChainId,
+  address proxyAdminOwner,
+  SystemConfigInputs memory systemConfigInputs,
+  L2OutputOracleInputs memory l2OutputOracleInputs
```

Comment:
It would be good to avoid enshrining anything about the L2OutputOracle in APIs.  It is deleted in the latest versions of contracts.  I know the plan is to start with MCP based on the above discussion but we shouldn't be introducing new APIs that are already broken/made redundant by the latest contract releases.

Reply:
The plan is actually to start with FPs now I need to update the spec. However, I wrote the spec to allow the inputs to the `deploy` method to change in future versions:

https://github.com/ethereum-optimism/specs/blob/9469d0a1bd0be4529e452fd079b84f49ab8ce783/specs/experimental/op-stack-manager.md#L255-L261

This is because our contracts are not in a state where we can guarantee no breaking changes to the required inputs. This means the options for the `deploy` method function signature e are:
1. The currently approach, i.e. strongly typed inputs but breaking changes to those types in future versions
2. A single `bytes` input so the actual signature stays consistent between versions, however the caller would still have to know how to encode the inputs, and this effectively has breaking changes anyway

Option (1) seemed more user friendly to me

Reply:
I definitely prefer option 1, smart contracts should not be calling this or they should be upgradable

Reply:
Another thought is to make the manager not upgradable and then deploy a new one for each release

Reply:
> Another thought is to make the manager not upgradable and then deploy a new one for each release

I've thought about this, the main downside is that it breaks down in a future where OPSM manages protocol upgrades—in that world, we likely need to make OPSM the ProxyAdmin owner (we can probably even remove the ProxyAdmin, but that's tangential for now), then the OPSM stores the mapping from chain -> owner address.

Reply:
Ah got it, that makes sense. Another possible world is that the OPSM isn't monolithic and responsible for every upgrade and instead there is a "migration contract" that you transfer ownership to and it migrates and then transfers ownership back. Then each contract is developed with one smaller purpose and we don't need infinite backwards compatibility, ie what if some chain doesn't upgrade on time and we update OPSM?

Reply:
> and we don't need infinite backwards compatibility, ie what if some chain doesn't upgrade on time and we update OPSM?

The key part here is that OPSM is focused on standard chains only, so if they are a standard chain they must upgrade on time by definition, and at some point we will begin enforcing that with atomic upgrades.

A non-proxied version where new ones are deployed in the future is definitely a valid design, however I think it would add more work in the long run and would prefer to keep this proxied. My rationale is as follows:
1. It's simpler from a docs, testing, and maintenance perspective when the address and source code file is static. For example, the superchain registry and doc address don't need to be continually updated, we don't need to carry around legacy test files or copy over test suites to new files.
2. The v2 of OPSM is unknown–it may be an update to change to deploy a new contract version, or it may be support for upgrades. Upgradeability keeps our options open, and we can always throw away the upgrade key
3. If/when OPSM does manage contract upgrades, it fits nicely into existing governance patterns—governance proposals now also include a new implementation address, and we upgrade OPSM as part of the governance proposal like we do with any other contract